### ISSUE TITLE
Fix context menu (Download URL, open in new window/tab)

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#688]
Right now we have `open link in new window` and `download link` in the link context menu which does not do the right thing.

We'd need to accommodate them for Next.

**Comment by Ambrevar** [comment]
Been working on this.  To add our own action to the WebKitGTK context menu, we need to create GActions.  See https://webkitgtk.org/reference/webkit2gtk/unstable/WebKitContextMenuItem.html.

Problem:

https://github.com/Ferada/cl-cffi-gtk/issues/41

In the mean time, the least we could do is remove the entries that are broken.  This does not require the creation of GActions.

**Comment by Ambrevar** [comment]
I've removed "paste" and "open in new window" with 86a873de20ba9289ec2d39ee3e1585dfa6ebd5fa.  Sadly "Download Linked File" persists, I don't know why.

**Comment by jmercouris** [comment]
Paste works for me... :-O

And now it is gone :'-(

Can you please put it in a `#-darwin` or so?

**Comment by Ambrevar** [comment]
Done!

**Comment by jmercouris** [comment]
Thank you!

**Comment by jmercouris** [comment]
This has been fixed on master. All of the context menus function as appropriately.

**Comment by Ambrevar** [comment]
Just tried it: right-click -> "Download Linked File" does not trigger
the download manager.

@jmercouris?

**Comment by Ambrevar** [comment]
Can you reproduce?

**Comment by jmercouris** [comment]
OK, so it download the file... but it does not trigger the download manager!

**Comment by jmercouris** [comment]
Fascinating, I wonder how it doesn't receive a signal for this. Perhaps there is yet another signal that we have to hook into. In any case, it /does/ work :-D

**Comment by Ambrevar** [comment]
I would say it does not, because it bypasses the handler, and thus it
ignores all the Nyxt options.  For instance, it downloads to $HOME
instead of the download folder.

It's probably trivial to fix but let's reopen to remember we need to fix this.

**Comment by Linden6** [comment]
The behaviour of the file manager not being called occurs for 'Save Image As' as well as 'Download Linked File'.  As with linked file, the image is saved to my home downloads folder, but no download manager is triggered.

For both actions the following message appears in the terminal:
(nyxt:5790): CRITICAL **: 17:55:27.648: WebKitContextMenuAction `webkit_context_menu_item_get_stock_action(WebKitContextMenuItem*): assertion 'WEBKIT_IS_CONTEXT_MENU_ITEM(item)' failed`

**Comment by Ambrevar** [comment]
Thanks for the report, this is helpful!

The freeze is probably fixed with the GI renderer.
@jmercouris Can you have a look at it?

**Comment by jmercouris** [comment]
Not sure if adding our own actions is the right way to go about this. I think we should just set the download location for the renderer based downloader.

**Comment by jmercouris** [comment]
That seems to be much more inline with what the APIs want us to do. A much more reliable/straightforwards path.

**Comment by jmercouris** [comment]
I've posted to the mailing list to see if someone knows why we do not get a "decide-policy" signal when right clicking and downloading.

**Comment by jmercouris** [comment]
It appears that the only way to do this would be to make our own GActions. I will investigate further. In the worst case we can remove the GAction for downloading.

**Comment by Ambrevar** [comment]
If I recall correctly, I had looked at GActions and could figure out how
to use them with cl-cffi-gtk.  Not even sure it's possible.

It could be worth it to look at Crategus' cl-cffi-gtk, it might have
updates with GAction examples.

Should cl-cffi-gtk be limiting here, we can try with cl-gobject-introspection.

**Comment by jmercouris** [comment]
OR we could just leave them a lone, and leave them with a disclaimer. They aren't broken, they just don't go through our download engine.

**Comment by Ambrevar** [comment]
It would feel very unfinished: no progress display, no listing, no way
to cancel the download, no error reporting, etc.

I'd rather remove it from the context menu.  After all, we support
hint downloads, so we are not losing any feature here.

**Comment by jmercouris** [comment]
I've now removed it from the context menu. I think in the future we could add renderer specific configuration options like this behind `#+features`. For example we could have it be disabled by default, but a user could add it into their `#+features` and then it would be enabled.

Why features? It is not a true configuration of Nyxt, it is an implementation detail of this renderer. It seems the right usage for features.

All that said, at least we fixed open in new tab! Perhaps in the future when we figure out why GAction creation is not working we can make our own custom GActions (which is what I imagine Epiphany/Gnome web/others must be doing).
